core/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java

```diff
@@ -1510,24 +1521,33 @@ public void restore() throws IOException {
                         index.totalRecoverFiles(), new ByteSizeValue(index.totalRecoverBytes()), index.reusedFileCount(), new ByteSizeValue(index.reusedFileCount()));
                 }
                 try {
-                    // first, delete pre-existing files in the store that have the same name but are
-                    // different (i.e. different length/checksum) from those being restored in the snapshot
-                    for (final StoreFileMetaData storeFileMetaData : diff.different) {
-                        IOUtils.deleteFiles(store.directory(), storeFileMetaData.name());
-                    }
+                    // list of all existing store files without the identical ones
+                    final Set<String> deleteIfExistFiles = Sets.difference(
+                        new HashSet<>(Arrays.asList(store.directory().listAll())),
+                        diff.identical.stream().map(StoreFileMetaData::name).collect(Collectors.toSet())
```

Comment:
Is there a need to substract the identical ones? The filesToRecover, which we iterate over, won't contain those anyhow.

Reply:
That was a safety check I added but I agree it does not make sense, I'll remove it.

Reply:
I agree with identical point. Could you also clean the trace logging 10 lines above while you are at it? I keep forgetting to do it and it doesn't make any sense now.

Also, can anyone think  of a scenario where case-insensitive FS can screw us here somehow?

Reply:
> Also, can anyone think of a scenario where case-insensitive FS can screw us here somehow?

I don't see any scenario like this...